### PR TITLE
AclMaintainer with dual stack and cfg/proxy container support

### PR DESCRIPTION
**Pull request by smorgrav**
The cfg/proxy support is basically about forwarding trusted ports as part of the ACL and apply the rules for IPv4 too.

After a lot of back and forth I ended up rewriting how we manipulate iptables, using iptables-restore instead of individual commands. One side-effect is that it is very easy to compare the actual iptables state to the wanted state instead of assuming the state as it is today, and we do everything in one command. 

Unit tests are stil in progress...

### CODE REVIEW

**Comment by smorgrav** [comment]
@freva This is work in progress but I would like to get your views on it before doing the last leg.

File: node-admin/src/main/java/com/yahoo/vespa/hosted/node/admin/maintenance/acl/AclMaintainer.java
Comment (freva) [inline node-admin/src/main/java/com/yahoo/vespa/hosted/node/admin/maintenance/acl/AclMaintainer.java]: This should probably throw an exception and the caller needs to know which IP addresses this host should have.

**Review by freva — COMMENTED** [review]
I haven't reviewed the `iptables` commands specifically, but the rest looks reasonable. 

Worth noting that this will double number of docker execs for ACL maintainer, which I think is the only piece of code that regularly does execs, so effectively this would almost double number of execs for `node-admin` as a whole. `exec` is quite expensive and there is an observed _correlation_ with number of execs/containers and various docker daemon crashes/hangs/errors.

**Comment by smorgrav** [comment]
@freva I don't think we are using docker exec here. We are simply executing the commands in an other network namespace using nsenter. Was that what you meant?

File: node-admin/src/main/java/com/yahoo/vespa/hosted/node/admin/maintenance/acl/AclMaintainer.java
Comment (smorgrav) [reply node-admin/src/main/java/com/yahoo/vespa/hosted/node/admin/maintenance/acl/AclMaintainer.java]: The normal case is where does not return an address is when the container does not have a ipv4 address (at least not DNS wise) which is the case for our tenant containers today. We could move that logic to the caller of course.

**Comment by freva** [comment]
@smorgrav Huh, that's interesting, the method you used is `dockerOperations.executeCommandInContainerAsRoot()` which does `docker exec`, but I guess you meant to use `dockerOperations.executeCommandInNetworkNamespace()`...? If so, then yes, there wont be any docker exec.

**Comment by smorgrav** [comment]
Ah... wrong execute cmd... thanks

**Comment by smorgrav** [comment]
@freva I think this is ready for a proper review. I'll test it live and verify the iptables rules against what we have on premise before I merge. I suspect I need to adjust a little bit on normalizing the rules.

**Comment by smorgrav** [comment]
@hakonhall FYI the changes to AclMaintainer

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/restapi/v2/NodeAclResponse.java
Comment (freva) [inline node-repository/src/main/java/com/yahoo/vespa/hosted/provision/restapi/v2/NodeAclResponse.java]: Looks like we call `nodeRepository.getNodeAcls(node, aclsForChildren)` 3 times now, call it once and reuse result?

File: node-admin/src/main/java/com/yahoo/vespa/hosted/node/admin/configserver/noderepository/RealNodeRepository.java
Comment (freva) [inline node-admin/src/main/java/com/yahoo/vespa/hosted/node/admin/configserver/noderepository/RealNodeRepository.java]: Why do we still have the second argument? Its not needed, and will this even work at all? This tries to match container names against `trustedBy`s, which is hostnames? So this will always return empty map?

File: node-admin/src/main/java/com/yahoo/vespa/hosted/node/admin/maintenance/acl/Acl.java
Comment (smorgrav) [reply node-admin/src/main/java/com/yahoo/vespa/hosted/node/admin/maintenance/acl/Acl.java]: Better - fixed

File: node-admin/src/main/java/com/yahoo/vespa/hosted/node/admin/maintenance/acl/Acl.java
Comment (smorgrav) [reply node-admin/src/main/java/com/yahoo/vespa/hosted/node/admin/maintenance/acl/Acl.java]: Fixed

File: node-admin/src/main/java/com/yahoo/vespa/hosted/node/admin/maintenance/acl/AclMaintainer.java
Comment (smorgrav) [reply node-admin/src/main/java/com/yahoo/vespa/hosted/node/admin/maintenance/acl/AclMaintainer.java]: Yeah... stupid default settings of IDEA

Comment (smorgrav) [reply node-admin/src/main/java/com/yahoo/vespa/hosted/node/admin/maintenance/acl/AclMaintainer.java]: Fixed

File: node-admin/src/main/java/com/yahoo/vespa/hosted/node/admin/maintenance/acl/AclMaintainer.java
Comment (smorgrav) [reply node-admin/src/main/java/com/yahoo/vespa/hosted/node/admin/maintenance/acl/AclMaintainer.java]: Fixed

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/restapi/v2/NodeAclResponse.java
Comment (smorgrav) [reply node-repository/src/main/java/com/yahoo/vespa/hosted/provision/restapi/v2/NodeAclResponse.java]: One time it is

File: node-admin/src/main/java/com/yahoo/vespa/hosted/node/admin/configserver/noderepository/RealNodeRepository.java
Comment (smorgrav) [reply node-admin/src/main/java/com/yahoo/vespa/hosted/node/admin/configserver/noderepository/RealNodeRepository.java]: Yeah, the original idea was to ask for individual acls as that made the logic on the caller side mush cleaner... but as that is not feasible I'll revert and add the hostname to the Acl class and make that class a little bit smarter with respect to net family presence (help solves the if address not present thing).

Comment (freva) [reply node-admin/src/main/java/com/yahoo/vespa/hosted/node/admin/configserver/noderepository/RealNodeRepository.java]: No, you can get the exact same return without sending the list of containers, see the long code snippet i wrote in this PR on April 9th.

Comment (smorgrav) [reply node-admin/src/main/java/com/yahoo/vespa/hosted/node/admin/configserver/noderepository/RealNodeRepository.java]: Well the point was to explicitly tell which ACL we wanted and not all... but changed this in later commit.

File: node-admin/src/main/java/com/yahoo/vespa/hosted/node/admin/maintenance/acl/Acl.java
Comment (freva) [inline node-admin/src/main/java/com/yahoo/vespa/hosted/node/admin/maintenance/acl/Acl.java]: First argument can be easily derived from the second

File: node-admin/src/main/java/com/yahoo/vespa/hosted/node/admin/configserver/noderepository/RealNodeRepository.java
Comment (freva) [inline node-admin/src/main/java/com/yahoo/vespa/hosted/node/admin/configserver/noderepository/RealNodeRepository.java]: Add `.distinct()` before `.forEach()` since there's likely overlap between the two key sets.

**Review by freva — APPROVED** [review]
Minor comments, LGTM

**Comment by smorgrav** [comment]
In addition to the unit tests, I've verified this to work for one aws porxy host. Will merge when green

**Comment by freva** [comment]
Also for a regular node in Yahoo?

**Comment by smorgrav** [comment]
Tested on premise too.

**Pull request merged**
